### PR TITLE
[faux-mgs] Make console less panic-happy

### LINKED ISSUE #122: faux-mgs panic while attached to scrimlet console
(Not certain if this is the correct repo; feel free to redirect as appropriate)

In today's dogfood update, we saw faux-mgs crash while connected to the console of a scrimlet:

Around when it crashed. Note the attempt to connect to the console again failed:
```
BRM42220051 # zpool list
NAME                                       SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
oxi_24d4ec6c-041a-4c79-abe9-c86316b68534   748G   644M   747G        -         -     0%     0%  1.00x    ONLINE  -
oxi_5a058adc-8208-4bff-b4f3-44e2651435b0   748G   644M   747G        -         -     0%     0%  1.00x    ONLINE  -
oxp_021afd19-2f87-4def-9284-ab7add1dd6ae  2.91T  46.3G  2.86T        -         -     0%     1%  1.00x    ONLINE  -
oxp_11b02ce7-7e50-486f-86c2-de8af9575a45  2.91T  13.8G  2.89T        -         -     0%     0%  1.00x    ONLINE  -
oxp_2d1ebe24-6deb-4f81-8450-6842de28126c  2.91T  63.4G  2.84T        -         -     1%     2%  1.00x    ONLINE  -
oxp_350b2814-7b7f-40f1-9bf6-9818a1ef49bb  2.91T  4.48M  2.91T        -         -     0%     0%  1.00x    ONLINE  -
oxp_4a624324-003a-4255-98e8-546a90b5b7fa  2.91T  31.5G  2.88T        -         -     0%     1%  1.00x    ONLINE  -
oxp_6b9ec5f1-859f-459c-9c06-6a51ba87786f  2.91T  35.0G  2.87T        -         -     0%     1%  1.00x    ONLINE  -
oxp_91ea7bb6-2be7-4498-9b0d-a0521509ec00  2.91T  39.3G  2.87T        -         -     0%     1%  1.00x    ONLINE  -
oxp_e12f29b8-1ab8-431e-bc96-1c1298947980  2.91T  9.21G  2.90T        -         -     0%     0%  1.00x    ONLINE  -
oxp_e6d2fe1d-c74d-40cd-8fae-bc7d06bdaac8  2.91T  70.3G  2.84T        -         -     0%     2%  1.00x    ONLINE  -
rpool                                     3.94G   773M  3.18G        -         -    11%    19%  1.00x    ONLINE  -
BRM42220051 # thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: RPC call failed (gave up after 5 attempts)', faux-mgs/src/usart.rs:86:58
                                                 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

Error: failed to send data (task shutdown?)

Caused by:
    channel closed
jordan@jeeves ~ $ pilot sp console BRM42220051
attaching to console; to detach, press: Ctrl-A, Ctrl-X...
 creating SP handle on interface rack2sw0tp1, component: faux-mgs
                                                                 Sep 07 18:33:19.382 INFO initial discovery complete, addr: [fe80::aa40:25ff:fe04:185%7]:11111, interface: rack2sw0tp1, component: faux-mgs
                                                                              Error: failed to attach to serial console

Caused by:
    0: Error response from SP: serial console already attached
    1: serial console already attached
```

Forcing a detach of the previous session allowed us to proceed on the console again:

```
jordan@jeeves ~ $ pilot sp console -D BRM42220051
attempting to detach console prior to attaching...
Ok("SP serial console detached\n")
attaching to console; to detach, press: Ctrl-A, Ctrl-X...
faux-mgs
        Sep 07 18:33:52.064 INFO initial discovery complete, addr: [fe80::aa40:25ff:fe04:185%7]:11111, interface: rack2sw0tp1, component: faux-mgs
```

### PR DESCRIPTION
**Pull request by jgallagher**
All the `.unwrap()`s in the `usart` module have been replaced. Most of them were fine and became `.expect("reason we think it's fine")`. The one that was not is in #122: if we fail a send to the SP, we previously panicked. We now buffer failed sends and will retry them (indefinitely, logging on each failure, except for a handful of errors we know to be fatal).

Example of talking to a gimletlet where I pull the network plug mid-session then restore it:

```
Sep 08 20:07:59.181 INFO creating SP handle on interface eth0, component: faux-mgs
Sep 08 20:07:59.183 INFO initial discovery complete, addr: [fe80::c1d:7dff:feef:9f1d%2]:11111, interface: eth0, component: faux-mgs
have I lost my connection?
Sep 08 20:08:18.767 WARN communication error with SP (will retry), err: RPC call failed (gave up after 5 attempts), component: faux-mgs
Sep 08 20:08:28.773 WARN communication error with SP (will retry), err: RPC call failed (gave up after 5 attempts), component: faux-mgs
Sep 08 20:08:34.779 INFO communication with SP reestablished, component: faux-mgs
the console is working again!
```

Example where we're attached to the console and someone else force-detaches us (faux-mgs exits after logging the error):

```
Sep 08 20:13:01.262 INFO creating SP handle on interface eth0, component: faux-mgs
Sep 08 20:13:01.264 INFO initial discovery complete, addr: [fe80::c1d:7dff:feef:9f1d%2]:11111, interface: eth0, component: faux-mgs
have I been detached?
Sep 08 20:13:14.480 ERRO fatal communication error with SP, err: Error response from SP: serial console is not attached, component: faux-mgs
```

Both of these examples result in panics on `main`.

Fixes #122.

### CODE REVIEW

File: faux-mgs/src/main.rs [hunk 1]
```diff
@@ -589,7 +590,17 @@ async fn main() -> Result<()> {
             .await?;
 
             // If usart::run() returns, the user detached; exit.
-            return Ok(());
+            //
+            // We don't just `return Ok(())` here because we'll bump into
```
Comment (andrewjstone) [inline faux-mgs/src/main.rs:594]: Nice workaround!

File: faux-mgs/src/usart.rs [hunk 2]
```diff
@@ -138,29 +163,87 @@ pub(crate) async fn run(
...
+) -> AttachedSerialConsoleSend {
     let mut keepalive = time::interval(SERIAL_CONSOLE_IDLE_TIMEOUT / 4);
     keepalive.set_missed_tick_behavior(MissedTickBehavior::Delay);
 
+    // If we fail to send a message, we need to resend it; we'll keep a running
+    // ticker and check for messages-to-send every second.
+    let mut check_for_resend = time::interval(Duration::from_secs(1));
+    check_for_resend.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+    let mut messages_to_send = VecDeque::new();
```
Comment (andrewjstone) [inline faux-mgs/src/usart.rs:208]: Presumably, this is only one or two messages long right?

**Review by andrewjstone — APPROVED** [review]

File: faux-mgs/src/usart.rs (hunk 2 above)
Comment (andrewjstone) [reply faux-mgs/src/usart.rs:208]: Got it. Still, human driven, and self imposed memory bloat based on text input. Seems fine!

File: faux-mgs/src/main.rs (hunk 1 above)
Comment (jgallagher) [reply faux-mgs/src/main.rs:594]: Credit to @dancrossnyc! I brought this up during last week's Rust study group, and he both immediately recognized the problem (before we found the relevant tokio issue) and offered the suggestion to explicitly exit.

**Pull request merged**